examples/experimental/native/flexible_primitives_mnist_pytorch/model_def.py

```diff
@@ -0,0 +1,199 @@
+"""
+This example shows how to interact with the Determined PyTorch interface to
+build a basic MNIST network.
+"""
+
+from typing import Any, Dict, Union, Sequence
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.optim.lr_scheduler import LambdaLR
+import torchvision
+from determined.tensorboard.metric_writers.pytorch import TorchWriter
+from determined.pytorch import PyTorchTrial, PyTorchTrialContext, DataLoader, LRScheduler
+
+import data
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+
+
+class Generator(nn.Module):
+    def __init__(self, latent_dim, img_shape):
+        super(Generator, self).__init__()
+        self.img_shape = img_shape
+
+        def block(in_feat, out_feat, normalize=True):
+            layers = [nn.Linear(in_feat, out_feat)]
+            if normalize:
+                layers.append(nn.BatchNorm1d(out_feat, 0.8))
+            layers.append(nn.LeakyReLU(0.2, inplace=True))
+            return layers
+
+        self.model = nn.Sequential(
+            *block(latent_dim, 128, normalize=False),
+            *block(128, 256),
+            *block(256, 512),
+            *block(512, 1024),
+            nn.Linear(1024, int(np.prod(img_shape))),
+            nn.Tanh()
+        )
+
+    def forward(self, z):
+        img = self.model(z)
+        img = img.view(img.size(0), *self.img_shape)
+        return img
+
+
+class Discriminator(nn.Module):
+    def __init__(self, img_shape):
+        super(Discriminator, self).__init__()
+
+        self.model = nn.Sequential(
+            nn.Linear(int(np.prod(img_shape)), 512),
+            nn.LeakyReLU(0.2, inplace=True),
+            nn.Linear(512, 256),
+            nn.LeakyReLU(0.2, inplace=True),
+            nn.Linear(256, 1),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, img):
+        img_flat = img.view(img.size(0), -1)
+        validity = self.model(img_flat)
+
+        return validity
+
+
+class GAN(nn.Module):
+    def __init__(self, trial_context: PyTorchTrialContext) -> None:
+        super(GAN, self).__init__()
+
+        self.context = trial_context
+
+        mnist_shape = (1, 28, 28)
+        self.generator = Generator(latent_dim=self.context.get_hparam("latent_dim"), img_shape=mnist_shape)
+        self.discriminator = Discriminator(img_shape=mnist_shape)
+
+    def forward(self, z):
+        return self.generator(z)
```

Comment:
non-blocking: Do people normally use submodels to implement GANs?  This "forward only passes through one submodel" behavior feels like we are introducing an unnatural layer of indirection to the example, especially when `train_batch()` uses `model.generator(z)` anyway.

I mark this as non-blocking because it is totally valid python code and I'm not actually sure if this is people prefer it for some other reason.

Reply:
Changed this example to using multiple models.

> This "forward only passes through one submodel" behavior feels like we are introducing an unnatural layer of indirection to the example, especially when train_batch() uses model.generator(z) anyway.

GAN does have better training performance in its discriminator than other discriminative methods. It aims at training a generative model so defining `forward` to return what generator forward returns totally makes sense.